torch/utils/data/dataloader.py

```diff
@@ -352,7 +352,11 @@ def _next_data(self):
 
     def __next__(self):
         data = self._next_data()
-        self._num_yielded += 1
+        # num_yielded and the value of _IterableDataset_len_called are in units of "samples" rather than
+        # batches, so when using auto_collation, we need to increment by the length of the batch we're
+        # yielding
+        self._num_yielded += len(data) if self._auto_collation else 1
```

Comment:
this is wrong through, because of collation. so what we can do is to gather the total #yields from workers and then also return them, maybe from `_next_data`.

Reply:
Is this wrong because data often something like a tuple or dict rather than a Tensor or custom object that can implement `__len__`?

Reply:
It is wrong because `data` will be the collated version. For example, if the data is of the form `(tensor_image, int_label)`, a collated batch of 10 samples will be a 2-tuple, `data = (batched_tensor_images, batched_tensor_labels)`, with each element having a leading dimension (batch dimension) of size 10. Yet `len(data) = 2`. Moreover, in general one can not detect number of samples from yielded data because of possible custom collate function. So this would require some deeper modification of the communication logic, and specifically including the workers communicating how many `next` they have called.